File: src/2048.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

// https://leetcode.com/problems/next-greater-numerically-balanced-number/solutions/1537631/next-permutation-vs-dfs/
int cnt[10] = {0, 1, 2, 3, 4, 5, 6, 7, 8, 9};
long long dfs(long long n, long long val, int sz) {
    long long retVal = 0;

    if (sz == 0) {
        for (int i = 1; i <= 9; ++i) {
            if ((cnt[i] != i) && (cnt[i] != 0)) {
                return retVal;
            }
        }
        retVal = (val > n) ? (val) : (0);

        return retVal;
    }

    for (int i = 1; retVal == 0 && i <= 9; ++i)
        if ((cnt[i] > 0) && (cnt[i] <= sz)) {
            --cnt[i];
            retVal = dfs(n, val * 10 + i, sz - 1);
            ++cnt[i];
        }

    return retVal;
}
int nextBeautifulNumber(int n) {
    int retVal = 0;

    char s[8];  // 0 <= n <= 10^6
    memset(s, 0, sizeof(s));
    snprintf(s, sizeof(s), "%d", n);
    int sz = strlen(s);
    retVal = dfs(n, 0, sz);
    if (retVal == 0) {
        retVal = dfs(0, 0, sz + 1);
    }

    return retVal;
}

int main(int argc, char** argv) {
    struct testCaseType {
        int n;
    } testCase[] = {{1}, {1000}, {3000}};
    int numberOfTestCase = sizeof(testCase) / sizeof(testCase[0]);
    /* Example
     *  Input: n = 1
     *  Output: 22
     *
     *  Input: n = 1000
     *  Output: 1333
     *
     *  Input: n = 3000
     *  Output: 3133
     */

    int answer;
    int i;
    for (i = 0; i < numberOfTestCase; ++i) {
        printf("Input: n = %d\n", testCase[i].n);

        answer = nextBeautifulNumber(testCase[i].n);
        printf("Output: %d\n", answer);

        printf("\n");
    }

    return EXIT_SUCCESS;
}

File: src/2048.py
import logging
import sys
from bisect import bisect_left, bisect_right
from collections import Counter, defaultdict, deque
from functools import cache, lru_cache
from heapq import heapify, heappop, heappush
from itertools import accumulate
from typing import List, Optional


def logging_setting():
    LOG_LEVEL = logging.INFO    # DEBUG, INFO, WARNING, ERROR, CRITICAL
    if __debug__:
        LOG_LEVEL = logging.DEBUG

    logging.basicConfig(
        stream=sys.stderr,
        level=LOG_LEVEL,
        format="%(levelname)-6s %(asctime)s %(module)s:%(funcName)s:%(lineno)-4s %(message)s",
        datefmt='%Y/%m/%d %H:%M:%S')


# https: // leetcode.com/problems/next-greater-numerically-balanced-number/solutions/1537631/next-permutation-vs-dfs/
class Solution:
    def __init__(self):
        self.cnt = [i for i in range(10)]

    def dfs(self, n: int, val: int, sz: int) -> int:
        retVal = 0

        if sz == 0:
            for i in range(1, 10):
                if self.cnt[i] != i and self.cnt[i] != 0:
                    return retVal
            retVal = val if val > n else 0

            return retVal

        for i in range(1, 10):
            if retVal == 0 and self.cnt[i] > 0 and self.cnt[i] <= sz:
                self.cnt[i] -= 1
                retVal = self.dfs(n, val * 10 + i, sz - 1)
                self.cnt[i] += 1

        return retVal

    def nextBeautifulNumber(self, n: int) -> int:
        retVal = 0

        sz = len(str(n))
        retVal = self.dfs(n, 0, sz)
        if retVal == 0:
            retVal = self.dfs(0, 0, sz + 1)

        return retVal


if __name__ == "__main__":
    logging_setting()

    try:
        logging.info("sys.version: %s", sys.version)
        print()

        pSolution = Solution()
        for n in [1, 1000, 3000]:
            # /* Example
            #  *  Input: n = 1
            #  *  Output: 22
            #  *
            #  *  Input: n = 1000
            #  *  Output: 1333
            #  *
            #  *  Input: n = 3000
            #  *  Output: 3133
            #  */
            logging.info("Input: n = %s", n)

            retVal = pSolution.nextBeautifulNumber(n)
            logging.info("Output: %s", retVal)

            print()
    except KeyboardInterrupt as exception:
        logging.error("%s: %s", exception.__class__.__name__, exception, exc_info=True)
        pass
    except Exception as exception:
        logging.error("%s: %s", exception.__class__.__name__, exception, exc_info=True)
        pass

    sys.exit(0)
